src/coins.cpp

```diff
@@ -573,96 +594,49 @@ bool CCoinsViewCache::HaveCswNullifier(const uint256& scId,
+            libzendoomc::CalculateCumulativeCertDataHash(prevEpochCumulativeCertDataHash, prevEpochCertDataHash, updatedprevEpochCumulativeCertDataHash);
+        } else {
+            // use default value prevEpochCumulativeCertDataHash
+            updatedprevEpochCumulativeCertDataHash = libzendoomc::ScFieldElement();
         }
 
-        std::pair<CCertDataHashMap::iterator, bool> ret = cacheCertDataHashes.insert(std::make_pair(position, CCertDataHashCacheEntry()));
-        ret.first->second.certDataHash = certDataHash;
-        ret.first->second.certDataCumulativeHash = certDataCumulativeHash;
-        ret.first->second.flag = CCertDataHashCacheEntry::FRESH;
+        CCertDataHashCacheEntry newEntry{std::make_pair(certDataHash, updatedprevEpochCumulativeCertDataHash),CCertDataHashCacheEntry::FRESH};
+        this->cacheCertDataHashes[std::make_pair(scId, epoch)] = newEntry;
     }
 }
 
-bool CCoinsViewCache::GetCertDataHash(const uint256& scId,
-                     const int epoch,
-                     libzendoomc::ScFieldElement& certDataHash) const {
-    std::pair<uint256, int> position = std::make_pair(scId, epoch);
-    
-    CCertDataHashMap::iterator it = cacheCertDataHashes.find(position);
-    if (it != cacheCertDataHashes.end()) {
-        certDataHash = it->second.certDataHash;
+bool CCoinsViewCache::HaveCertDataHashes(const uint256& scId, const int epoch) const {
+    CCertDataHashMap::const_iterator it = FetchCertDataEntry(scId, epoch);
+    if (it != cacheCertDataHashes.end() && it->second.flag != CCertDataHashCacheEntry::Flags::ERASED)
         return true;
-    }
-
-    libzendoomc::ScFieldElement tmp;
-    bool res = base->GetCertDataHash(scId, epoch, tmp);
-
-    if (res) {
-        certDataHash = tmp;
-        
-        std::pair<CCertDataHashMap::iterator, bool> ret = cacheCertDataHashes.insert(std::make_pair(position, CCertDataHashCacheEntry()));
-        ret.first->second.flag = CCertDataHashCacheEntry::DEFAULT;
-        ret.first->second.certDataHash = tmp;
-        base->GetCertDataCumulativeHash(scId, epoch, tmp);
-        ret.first->second.certDataCumulativeHash = tmp;
-    }
 
-    return res;
+    return false;
 }
 
-bool CCoinsViewCache::GetCertDataCumulativeHash(const uint256& scId,
-                                                const int epoch,
-                                                libzendoomc::ScFieldElement& certDataCumulativeHash) const {
-
-    std::pair<uint256, int> position = std::make_pair(scId, epoch);
-
-    CCertDataHashMap::iterator it = cacheCertDataHashes.find(position);
-    if (it != cacheCertDataHashes.end()) {
-        certDataCumulativeHash = it->second.certDataCumulativeHash;
+bool CCoinsViewCache::GetCertDataHashes(const uint256& scId, const int epoch,
+                       std::pair<libzendoomc::ScFieldElement, libzendoomc::ScFieldElement>& certDataHashes) const
+{
+    CCertDataHashMap::const_iterator it = FetchCertDataEntry(scId, epoch);
```

Comment:
We also have to use this function in RemoveCertDataHash.

Reply:
Agree, done